Mathlib/Topology/Algebra/Module/Action.lean

```diff
@@ -0,0 +1,253 @@
+  /- Proof: We need to prove that the product topology is finer than the pullback
+     of the action topology. But the action topology is an Inf and thus a limit,
+     and pullback is a right adjoint, so it preserves limits.
+     We must thus show that the product topology is finer than an Inf, so it suffices
+     to show it's a lower bound, which is not hard. All this is wrapped into
+     `continuousSMul_sInf`.
+  -/
+  continuousSMul_sInf <| fun _ _ ↦ by simp_all only [Set.mem_setOf_eq]
+
+/-- Addition `+ : A × A → A` is continuous if `R` is a topological
+ring, and `A` is an `R` module with the action topology. -/
+theorem ActionTopology.continuousAdd : @ContinuousAdd A (actionTopology R A) _ :=
+  continuousAdd_sInf <| fun _ _ ↦ by simp_all only [Set.mem_setOf_eq]
+
+instance IsActionTopology.toContinuousSMul [TopologicalSpace A] [IsActionTopology R A] :
+    ContinuousSMul R A := eq_ActionTopology R A ▸ ActionTopology.continuousSMul R A
+
+-- this can't be an instance because typclass inference can't be expected to find `R`.
+theorem IsActionTopology.toContinuousAdd [TopologicalSpace A] [IsActionTopology R A] :
+    ContinuousAdd A := eq_ActionTopology R A ▸ ActionTopology.continuousAdd R A
+
+/-- The action topology is `≤` any topology making `A` into a topological module. -/
+theorem actionTopology_le [τA : TopologicalSpace A] [ContinuousSMul R A] [ContinuousAdd A] :
+    actionTopology R A ≤ τA := sInf_le ⟨inferInstance, inferInstance⟩
+
+end basics
+
+namespace IsActionTopology
+
+section subsingleton
+
+instance instSubsingleton (R : Type*) [TopologicalSpace R] (A : Type*) [Add A] [SMul R A]
+    [Subsingleton A] [TopologicalSpace A] : IsActionTopology R A where
+  eq_actionTopology' := by
+    ext U
+    simp only [isOpen_discrete]
+
+end subsingleton
+
+section self
+variable (R : Type*) [Semiring R] [τR : TopologicalSpace R] [TopologicalSemiring R]
+
+/-!
+We now fix once and for all a topological semiring `R`.
+
+We first prove that the action topology on `R` considered as a module over itself,
+is `R`'s topology.
+-/
+
+/-- The action by left-multiplication is the action topology. -/
+instance _root_.TopologicalSemiring.toIsActionTopology : IsActionTopology R R where
+  eq_actionTopology' := by
+    /- Let `R` be a topological ring with topology τR. To prove that `τR` is the action
+    topology, it suffices to prove inclusions in both directions.
+    One way is obvious: addition and multiplication are continuous for `R`, so the
+    action topology is finer than `R` because it's the finest topology with these properties.-/
+    refine le_antisymm ?_ (actionTopology_le R R)
+    /- The other way is more interesting. We can interpret the problem as proving that
+    the identity function is continuous from `R` with the action topology to `R` with
+    its usual topology to `R` with the action topology.
+    -/
+    rw [← continuous_id_iff_le]
```

Comment:
I think this should be the definition; namely, `IsActionTopology` should mean "the canonical map from `WithActionTopology R` to `R` is continuous", where `WithActionTopology R` is a type synonym that puts the action topology on `R`.

This should eliminate all the `@`s below!

Reply:
I tried this out in eric-wieser/withActionTopology, but there's a little more boilerplate needed than expected.

Reply:
Are you suggesting I should refactor the definition in this PR?

Reply:
After a private discussion with Eric he suggested that rather than the refactor I should add the lemma saying that if A is a topological module and the identity map from (A,standard) to (A,module topology) is continuous then A has the module topology. I've added this in `ad2d292`.